WIP: Update IGeometryGrid API for set/get Dimension/Origin/Spacing

imikejackson opened the pull request:
The API is changed to use a FloatVec3Type or SizeVec3Type (both alias for SIMPLArray<T, N>) so that
the API is consistent among the various classes and easy to use within SIMPL/DREAM.3D and its plugins.
The use of the std::tuple<T> based return types is deprecated and removed.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Update IGeometryGrid API for set/get Dimension/Origin/Spacing

The API is changed to use a FloatVec3Type or SizeVec3Type (both alias for SIMPLArray<T, N>) so that
the API is consistent among the various classes and easy to use within SIMPL/DREAM.3D and its plugins.
The use of the std::tuple<T> based return types is deprecated and removed.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson commented:
All of the other repos are also updated with the same branch. I am just putting in this PR rather than a PR for all 20 repositories.

I also added a few more methods to the SIMPLArray subclasses to convert to another container type (such as QVector or std::vector) and convert to SIMPLArray of another data type.

mmarineBlueQuartz reviewed (approved):
This looks good, but it might be worth it to mention that toContainer() requires the ContainerType to allow random access in the `@brief` description.  Many STL containers do not fit this requirement.

joeykleingers reviewed (approved):
Looks good.

imikejackson committed:
Fix failing unit test for ImageGeometry.

It was always failing due to an incorrect call to setSpacing instead of getSpacing
Also updated the ImageGeom API with more const correctness based on clang-tidy feedback

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson commented:
@mmarineBlueQuartz True. Maybe I should just use the most general case for the concrete implementations. I could probably use an initializer list since we know the size of the array in those functions. I think most (all?) the STL containers and QVector can handle an initializer list.

imikejackson commented:
Also thought about implementing some stream operators so we can have debug output of the class.

imikejackson committed:
Add robust toContainer() method, Update ImageGeom info string.

Add static method to IGeometry to generate a QVector<QString> of all the length Units

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson commented:
I added the requested changes to SIMPLArray, added some more constructors. Update the Doxygen for SIMPLArray to state which container types work. Updated the unit test for SIMPLArray to add the Vec4 and Vec6 cases, and add instantiations of the toContainer for std::vector, std::list, std::deque, QVector and QList.

mmarineBlueQuartz commented on Source/SIMPLib/Common/SIMPLArray.hpp:
```diff
@@ -172,6 +184,22 @@ class SIMPLArray
+  template <typename OutContainerType>
+  OutContainerType toContainer()
+  {
+    OutContainerType dest;
```
If you don't mind losing QList (which is suggested to avoid like the plague), pass in the dimentions to `OutContainerType dest(m_Array.size());` and keep your [] operators for assignment and copying.  Every other container mentioned in the description allows that pattern, and the most likely conversions (std::vector and QVector) do so optimally.  std::list and std::deque are slower when it comes to random access but are similarly unlikely conversions from an array.

imikejackson replied:
All requests implemented.

imikejackson committed:
Use operator[] in the toContainer() method. Fix failing unit test.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson commented:
This PR is superseded by PR #345
